ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java

```diff
@@ -240,42 +247,58 @@ public void receiveData(byte[] data) {
-     * The response is written to the Port named "output"
+     * Creating the string, which will be sent as a TCP package. See API and possible commands at
+     * http://www.irtrans.de/download/Docs/IRTrans%20TCP%20ASCII%20Schnittstelle_DE.pdf The response is written to the Port named "output"
      *
      * @param cmdString
-     *            The string to be sent, must not contain the prefix "A", it is
-     *            appended in the method in order to be compatible to old
-     *            UDP-commands
+     *            The string to be sent, must not contain the prefix "A", it is appended in the method in order to be compatible to old UDP-commands
      */
     private void sendString(String cmdString) {
         try {
+            tcpSocket = initTcpSocket(tcpSocket);
             int readTimeout = cmdString.contains("learn") ? READ_TIMEOUT_LEARN_MS : READ_TIMEOUT_SEND_MS;
-            if (socket == null || socket.isClosed() || !socket.isConnected()) {
-                initSocket();
-            }
-            if (socket != null) {
-                sendToSocket(cmdString);
-                String result = readFromSocket(readTimeout);
+            if (tcpSocket != null) {
+                sendToSocket(tcpSocket, cmdString);
+                String result = readFromSocket(tcpSocket, readTimeout);
                 AstericsErrorHandling.instance.reportInfo(this, "IRTrans response: " + result);
-                opOutput.sendData(result.getBytes());
+                opOutputResult.sendData(result.getBytes());
             } else {
-                opOutput.sendData(ERROR_RESULT.getBytes());
+                opOutputResult.sendData(ERROR_RESULT.getBytes());
             }
         } catch (Exception e) {
-            opOutput.sendData(ERROR_RESULT.getBytes());
-            closeSocket();
+            opOutputResult.sendData(ERROR_RESULT.getBytes());
+            closeTcpSocket(tcpSocket);
             AstericsErrorHandling.instance.reportError(this,
                     "The IRTrans - Plugin could not send data. Please verify that the IRTrans Module is connected and installed, and that the IP-address is correctly specified.");
             AstericsErrorHandling.instance.reportInfo(this, e.toString());
         }
     }
 
+    private void receiveCommandsToOutputPort() {
+        readerThread = new Thread(new Runnable() {
+
+            @Override
+            public void run() {
+                try {
+                    tcpSocketRead = initTcpSocket(tcpSocketRead);
+                    while (!readerThread.isInterrupted()) {
+                        String result = readFromSocket(tcpSocketRead, 0);
+                        opOutput.sendData(ConversionUtils.stringToBytes(result));
+                        logger.log(Level.INFO, "IRTrans received: " + result);
+                    }
+                } catch (Exception e) {
+                    logger.log(Level.INFO, "reading from socket to receive command stopped");
+                }
```

Comment:
closing the tcpSocketRead should be done in a finally block here, to ensure that it is closed both in case of exception and also after normal interruption.

finally {
     if(tcpSocketRead !=null) tcpSocketRead.close();
}